### ISSUE TITLE
Issue with doctrine/orm 2.12.x

### ISSUE THREAD

**Issue by michnovka** [symfony/maker-bundle#1090]
### Symfony version(s) affected

5.4.7

### Description

Symfony make entity command results in Exception

`symfony console make:entity`

`[critical] Error thrown while running command "make:entity". Message: "Cannot access protected property Doctrine\ORM\Mapping\Driver\AttributeDriver::$classNames"`

### How to reproduce

1. Set up new Symfony project
2. Use doctrine/orm 2.12.x-dev package
3. run make:entity

**Comment by stof** [comment]
I transferred the issue to MakerBundle, which is the proper place to report that.

**Comment by michnovka** [comment]
Doctrine ORM 2.12.x is released and this issue persists.

**Comment by michnovka** [comment]
The issue stems from https://github.com/symfony/maker-bundle/blob/c5575532031eab8592ab6b1226e893b1a6b5bb0e/src/Doctrine/DoctrineHelper.php#L180-L187

I am not sure whats going on here. The `ReflectionClass` uses `Doctrine\Persistence\Mapping\Driver\AnnotationDriver` But in my case `DoctrineHelper::mappingDriversByPrefix` contains instance of `Doctrine\ORM\Mapping\Driver\AttributeDriver`.

So we are mixing types in here. Why one is from ORM, one is from Persistance?

the fix for my specific case is to simply swap

```php
$classNames = (new \ReflectionClass(AnnotationDriver::class))->getProperty('classNames');
```
for
```php
$classNames = (new \ReflectionClass(\Doctrine\ORM\Mapping\Driver\AttributeDriver::class))->getProperty('classNames');
```

But I doubt this will be the generic solution we are looking for.

**Comment by michnovka** [comment]
FYI there was Doctrine/Persistance 3.0 released, which might be the real cause of the issue, since doctrine/orm 2.12.x uses it

**Comment by RikudouSage** [comment]
Any workaround for this?

**Comment by kostyan-org** [comment]
found the same problem today

_"symfony/maker-bundle": "^1.38"
"doctrine/orm": "^2.12"
"doctrine/doctrine-bundle": "^2.6"_

symfony console make:entity

In DoctrineHelper.php line 187:
  Cannot access protected property Doctrine\ORM\Mapping\Driver\AnnotationDriver::$classNames

**Comment by kostyan-org** [comment]
"symfony/maker-bundle": "1.28.0"
version without bug

**Comment by Bechir-Marco** [comment]
I have just solved this issue by modifying "doctrine/annotations "version  in composer.json after checking which versions are allowed in composer.lock "conflict"

**Comment by lhapaipai** [comment]
> I have just solved this issue by modifying "doctrine/annotations "version in composer.json after checking which versions are allowed in composer.lock "conflict"

Hi @Bechir-Marco ,
thanks for your feedback, can you show us your `composer.json` file ?

**Comment by Bechir-Marco** [comment]
`{
    "type": "project",
    "license": "proprietary",
    "require": {
        "php": ">=7.4",
        "ext-ctype": "*",
        "ext-iconv": "*",
        "doctrine/annotations": "1.12.1",
        "doctrine/doctrine-bundle": "^2.2",
        "doctrine/doctrine-migrations-bundle": "^3.0",
        "doctrine/orm": "2.8",
        "gesdinet/jwt-refresh-token-bundle": "^1.1",
        "phpdocumentor/reflection-docblock": "^5.3",
        "sensio/framework-extra-bundle": "^6.2",
        "stof/doctrine-extensions-bundle": "^1.7",
        "symfony/asset": "4.4.*",
        "symfony/console": "4.4.*",
        "symfony/dotenv": "4.4.*",
        "symfony/expression-language": "4.4.*",
        "symfony/flex": "^1.3.1",
        "symfony/form": "4.4.*",
        "symfony/framework-bundle": "4.4.*",
        "symfony/http-client": "4.4.*",
        "symfony/intl": "4.4.*",
        "symfony/mailer": "4.4.*",
        "symfony/monolog-bundle": "^3.1",
        "symfony/process": "4.4.*",
        "symfony/property-access": "4.4.*",
        "symfony/property-info": "4.4.*",
        "symfony/proxy-manager-bridge": "4.4.*",
        "symfony/security-bundle": "4.4",
        "symfony/serializer": "4.4.*",
        "symfony/translation": "4.4.*",
        "symfony/twig-bundle": "4.4.*",
        "symfony/validator": "4.4.*",
        "symfony/web-link": "4.4.*",
        "symfony/web-server-bundle": "4.4.*",
        "symfony/yaml": "4.4.*",
        "symfonycasts/verify-email-bundle": "^1.10",
        "twig/extra-bundle": "^2.12|^3.0",
        "twig/twig": "^2.12|^3.0",
        "vich/uploader-bundle": "1.19.0"
    },
    "require-dev": {
        "phpunit/phpunit": "^9.5",
        "symfony/browser-kit": "4.4.*",
        "symfony/css-selector": "4.4.*",
        "symfony/debug-bundle": "4.4.*",
        "symfony/maker-bundle": "v1.30.1",
        "symfony/phpunit-bridge": "^6.0",
        "symfony/stopwatch": "4.4.*",
        "symfony/web-profiler-bundle": "4.4.*"
    },
    "config": {
        "allow-plugins": {
            "composer/package-versions-deprecated": true,
            "symfony/flex": true
        },
        "preferred-install": {
            "*": "dist"
        },
        "sort-packages": true
    },
    "autoload": {
        "psr-4": {
            "App\\": "src/"
        }
    },
    "autoload-dev": {
        "psr-4": {
            "App\\Tests\\": "tests/"
        }
    },
    "replace": {
        "paragonie/random_compat": "2.*",
        "symfony/polyfill-ctype": "*",
        "symfony/polyfill-iconv": "*",
        "symfony/polyfill-php71": "*",
        "symfony/polyfill-php70": "*",
        "symfony/polyfill-php56": "*"
    },
    "scripts": {
        "auto-scripts": {
            "cache:clear": "symfony-cmd",
            "assets:install %PUBLIC_DIR%": "symfony-cmd"
        },
        "post-install-cmd": [
            "@auto-scripts"
        ],
        "post-update-cmd": [
            "@auto-scripts"
        ]
    },
    "conflict": {
        "symfony/symfony": "*"
    },
    "extra": {
        "symfony": {
            "allow-contrib": false,
            "require": "4.4.*"
        }
    }
}
`

**Comment by snipershady** [comment]
> found the same problem today
> 
> _"symfony/maker-bundle": "^1.38" "doctrine/orm": "^2.12" "doctrine/doctrine-bundle": "^2.6"_
> 
> symfony console make:entity
> 
> In DoctrineHelper.php line 187: Cannot access protected property Doctrine\ORM\Mapping\Driver\AnnotationDriver::$classNames

same problem on a brand new project symfony 6.0

**Comment by MatuxGG** [comment]
Same issue

**Comment by czachor** [comment]
Temporary workaround

Composer.json - change:

    "doctrine/orm": "^2.12",

to:

    "doctrine/orm": "^2.11",

add (to prevent 2.12 being installed):

    "conflict": {
      "symfony/symfony": "*",
      "doctrine/orm": "2.12.0"
    },

**Comment by Amand04** [comment]
Même problème que Tireur d'Elite. 
J'ai suivi les indications de czachor, j'ai modifié le fichier composer.json.
J'ai réussi quand même à créer mon entité et son repository en donnant en nom à l'entité dans l'instruction. Il m'a quand même indiqué:

Dans DoctrineHelper.php ligne 187 : Impossible d'accéder à la propriété protégée Doctrine\ORM\Mapping\Driver\AnnotationDriver::$classNames

mais l'a crée malgrés tout...?!

**Comment by Th3BiSh0p** [comment]
> Même problème que Tireur d'Elite. J'ai suivi les indications de czachor, j'ai modifié le fichier composer.json. J'ai réussi quand même à créer mon entité et son repository en donnant en nom à l'entité dans l'instruction. Il m'a quand même indiqué:
> 
> Dans DoctrineHelper.php ligne 187 : Impossible d'accéder à la propriété protégée Doctrine\ORM\Mapping\Driver\AnnotationDriver::$classNames
> 
> mais l'a crée malgrés tout...?!

tu as fait un composer update après la modification du composer.json ? Pour moi ça a bien contourné le problème.

**Comment by mbabker** [comment]
The issue is not the ORM 2.12 release.

The root issue is this bundle is not compatible with `doctrine/persistence` 3.0, and this bundle does not declare a dependency to `doctrine/persistence` to lock it to supported versions, so through transient dependencies an incompatible set of libraries is being installed.  Lock your application to `doctrine/persistence:2.*` and things should work until this bundle supports the new major version of the Persistence library.

(On a side note, I really don't see how anyone's getting an error from the `Doctrine\ORM\Mapping\Driver\AnnotationDriver` class because the `DoctrineHelper` is using the deprecated `Doctrine\Persistence\Mapping\Driver\AnnotationDriver`)

**Comment by mbabker** [comment]
See https://github.com/doctrine/orm/blob/2.12.x/UPGRADE.md#bc-break-attributedriver-and-annotationdriver-no-longer-extends-parent-class-from-doctrinepersistence

The `$classNames` property being reflected is on a parent class that is no longer inherited from.  This crude patch should help make things work while someone makes a "better" fix (I generally don't use this bundle, so I can't say if it's any good):

```patch
diff --git a/src/Doctrine/DoctrineHelper.php b/src/Doctrine/DoctrineHelper.php
index 8c05db2..48cdad6 100644
--- a/src/Doctrine/DoctrineHelper.php
+++ b/src/Doctrine/DoctrineHelper.php
@@ -177,14 +177,17 @@ final class DoctrineHelper
      */
     public function getMetadata(string $classOrNamespace = null, bool $disconnected = false)
     {
-        $classNames = (new \ReflectionClass(AnnotationDriver::class))->getProperty('classNames');
-        $classNames->setAccessible(true);
-
         // Invalidating the cached AnnotationDriver::$classNames to find new Entity classes
         foreach ($this->mappingDriversByPrefix ?? [] as $managerName => $prefixes) {
             foreach ($prefixes as [$prefix, $annotationDriver]) {
                 if (null !== $annotationDriver) {
-                    $classNames->setValue($annotationDriver, null);
+                    $refl = new \ReflectionClass($annotationDriver);
+
+                    if ($refl->hasProperty('classNames')) {
+                        $classNames = (new \ReflectionClass($annotationDriver))->getProperty('classNames');
+                        $classNames->setAccessible(true);
+                        $classNames->setValue($annotationDriver, null);
+                    }
                 }
             }
         }
@@ -218,7 +221,11 @@ final class DoctrineHelper
                     $metadataDriver = $em->getConfiguration()->getMetadataDriverImpl();
                     if ($this->isInstanceOf($metadataDriver, MappingDriverChain::class)) {
                         foreach ($metadataDriver->getDrivers() as $driver) {
-                            if ($this->isInstanceOf($driver, AnnotationDriver::class)) {
+                            $refl = new \ReflectionClass($driver);
+
+                            if ($refl->hasProperty('classNames')) {
+                                $classNames = (new \ReflectionClass($driver))->getProperty('classNames');
+                                $classNames->setAccessible(true);
                                 $classNames->setValue($driver, null);
                             }
                         }
```

**Comment by lionelkimbs** [comment]
This worked for me :
1. Edit `composer.json` to change `doctrine/orm` version
```
...
    "require": {
        ...
        "doctrine/orm": "^2.11",
        ...
    }
    ...
    "conflict": {
        "symfony/symfony": "*",
        "doctrine/orm": "2.12.0"
    },
...
```
2. Update dependencies : `composer update`
3. Enjoy !

**Comment by kacperjcb** [comment]
This worked for me: 
composer require doctrine/orm:2.11

**Comment by sinE800** [comment]
> Temporary workaround
> 
> Composer.json - change:
> 
> ```
> "doctrine/orm": "^2.12",
> ```
> 
> to:
> 
> ```
> "doctrine/orm": "^2.11",
> ```
> 
> add (to prevent 2.12 being installed):
> 
> ```
> "conflict": {
>   "symfony/symfony": "*",
>   "doctrine/orm": "2.12.0"
> },
> ```

Merci/thanks.
(I hope they fix the 2.12 version)

**Comment by Amand04** [comment]
> > Même problème que Tireur d'Elite. J'ai suivi les indications de czachor, j'ai modifié le fichier composer.json. J'ai réussi quand même à créer mon entité et son référentiel en donnant en nom à l'entité dans l'instruction. Il m'a quand même indiqué :
> > Dans DoctrineHelper.php ligne 187 : Impossible d'accéder à la propriété protégée Doctrine\ORM\Mapping\Driver\AnnotationDriver::$classNames
> > mais l'a crée malgrés tout...?!
> 
> tu as fait un composer update après la modification du composer.json ? Pour moi ça a bien contourné le problème.

Oui cela fonctionne. Merci pour l'indication je suis une grande débutante...

**Comment by kanuragan** [comment]
> Temporary workaround
> 
> Composer.json - change:
> 
> ```
> "doctrine/orm": "^2.12",
> ```
> 
> to:
> 
> ```
> "doctrine/orm": "^2.11",
> ```
> 
> add (to prevent 2.12 being installed):
> 
> ```
> "conflict": {
>   "symfony/symfony": "*",
>   "doctrine/orm": "2.12.0"
> },
> ```

**Comment by kanuragan** [comment]
thanks, it worked

**Comment by snipershady** [comment]
downgrade works fine, but I hope doctrine developers will release a new version soon.
It's really unpleasant to run a brand new project and receive an exception from a "core" package.

**Comment by tourze** [comment]
<img width="369" alt="image" src="https://user-images.githubusercontent.com/13899502/164411459-fc535393-a8b6-4493-959a-0b88b06cdae5.png">
same problem

**Comment by stof** [comment]
Well, the issue is not in doctrine but in MakerBundle. When you use Reflection to access internal parts of a class, you cannot expect it to be covered by BC.

**Comment by snipershady** [comment]
maker bundle was update yesterday

` v1.39.0 2022-04-21 18:16 UTC`

but it still doesn't works !!!
@fabpot @nicolas-grekas  
I've  forwarded this issue to doctrine/orm here: https://github.com/doctrine/orm/issues/9669
but reading the answer from [greg0ire](https://github.com/greg0ire)
https://github.com/doctrine/orm/issues/9669#issuecomment-1105199656
looks like there is no communication between Symfony staff and doctrine maintainers


` php bin/console make:entity --verbose`

```
In DoctrineHelper.php line 187:
                                                                                              
  [Error]                                                                                     
  Cannot access protected property Doctrine\ORM\Mapping\Driver\AnnotationDriver::$classNames  
```
                                                                                              
```

Exception trace:
  at /home/shady/public_html/MY_NEW_PROJECT/vendor/symfony/maker-bundle/src/Doctrine/DoctrineHelper.php:187
 ReflectionProperty->setValue() at /home/shady/public_html/MY_NEW_PROJECT/vendor/symfony/maker-bundle/src/Doctrine/DoctrineHelper.php:187
 Symfony\Bundle\MakerBundle\Doctrine\DoctrineHelper->getMetadata() at /home/shady/public_html/MY_NEW_PROJECT/vendor/symfony/maker-bundle/src/Doctrine/DoctrineHelper.php:162
 Symfony\Bundle\MakerBundle\Doctrine\DoctrineHelper->getEntitiesForAutocomplete() at /home/shady/public_html/MY_NEW_PROJECT/vendor/symfony/maker-bundle/src/Maker/MakeEntity.php:515
 Symfony\Bundle\MakerBundle\Maker\MakeEntity->createEntityClassQuestion() at /home/shady/public_html/MY_NEW_PROJECT/vendor/symfony/maker-bundle/src/Maker/MakeEntity.php:116
 Symfony\Bundle\MakerBundle\Maker\MakeEntity->interact() at /home/shady/public_html/MY_NEW_PROJECT/vendor/symfony/maker-bundle/src/Command/MakerCommand.php:99
 Symfony\Bundle\MakerBundle\Command\MakerCommand->interact() at /home/shady/public_html/MY_NEW_PROJECT/vendor/symfony/console/Command/Command.php:276
 Symfony\Component\Console\Command\Command->run() at /home/shady/public_html/MY_NEW_PROJECT/vendor/symfony/console/Application.php:1007
 Symfony\Component\Console\Application->doRunCommand() at /home/shady/public_html/MY_NEW_PROJECT/vendor/symfony/framework-bundle/Console/Application.php:94
 Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() at /home/shady/public_html/MY_NEW_PROJECT/vendor/symfony/console/Application.php:299
 Symfony\Component\Console\Application->doRun() at /home/shady/public_html/MY_NEW_PROJECT/vendor/symfony/framework-bundle/Console/Application.php:80
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /home/shady/public_html/MY_NEW_PROJECT/vendor/symfony/console/Application.php:171
 Symfony\Component\Console\Application->run() at /home/shady/public_html/MY_NEW_PROJECT/vendor/symfony/runtime/Runner/Symfony/ConsoleApplicationRunner.php:54
 Symfony\Component\Runtime\Runner\Symfony\ConsoleApplicationRunner->run() at /home/shady/public_html/MY_NEW_PROJECT/vendor/autoload_runtime.php:29
 require_once() at /home/shady/public_html/MY_NEW_PROJECT/bin/console:11

make:entity [-a|--api-resource] [-b|--broadcast] [--regenerate] [--overwrite] [--] [<name>]
```

**Comment by kevinblewusi** [comment]
> This worked for me :
> 
> 1. Edit `composer.json` to change `doctrine/orm` version
> 
> ```
> ...
>     "require": {
>         ...
>         "doctrine/orm": "^2.11",
>         ...
>     }
>     ...
>     "conflict": {
>         "symfony/symfony": "*",
>         "doctrine/orm": "2.12.0"
>     },
> ...
> ```
> 
> 2. Update dependencies : `composer update`
> 3. Enjoy !

**Thanks, this work fine for me !**

**Comment by stof** [comment]
Well, the release of MakerBundle of yesterday fixes other things, but not that bug. Nobody has opened a PR with a fix yet.

**Comment by jrushlow** [comment]
A fix is in the works for this in #1095 or possibly a subsequent PR

**Comment by sarim** [comment]
While a proper fix is being worked on, maybe e hotfix release could be released with makerbundle's composer.json adding a conflict with doctrine/orm": "2.12.0"
